### ISSUE TITLE
macOS file change api can support file names as of 10.7

### ISSUE THREAD

**Issue by erg** [factor/factor#2066]
Investigate other platforms as well.

https://stackoverflow.com/questions/12693208/is-it-possible-to-get-the-filename-from-an-fsevent-id/12776403

```factor
[ "resource:files" t [ [ dup next-change . flush ] forever ] with-monitor ] with-monitors

! touch files/a
! should show file name but only shows directory
T{ file-change
    { path "/Users/erg/factor-master/files/" }
    { changed { +modify-file+ } }
    { monitor
        T{ macosx-monitor
            { path "/Users/erg/factor-master/files" }
            { queue T{ mailbox } }
            { handle
                T{ event-stream
                    { info ALIEN: 12 }
                    { handle ALIEN: 7fd610841380 }
                }
            }
        }
    }
}
```

**Comment by mrjbq7** [comment]
This might get close to it:

```diff
diff --git a/basis/io/monitors/macosx/macosx.factor b/basis/io/monitors/macosx/macosx.factor
index 75211a06b1..344e3fa640 100644
--- a/basis/io/monitors/macosx/macosx.factor
+++ b/basis/io/monitors/macosx/macosx.factor
@@ -1,19 +1,26 @@
 ! Copyright (C) 2008, 2010 Slava Pestov.
 ! See http://factorcode.org/license.txt for BSD license.
-USING: accessors arrays core-foundation.fsevents destructors fry
-io.backend io.monitors kernel locals sequences system ;
+USING: accessors arrays combinators core-foundation.fsevents
+destructors fry io.backend io.monitors kernel locals math
+sequences system ;
 IN: io.monitors.macosx
 
 TUPLE: macosx-monitor < monitor handle ;
 
 : enqueue-notifications ( triples monitor -- )
-    '[ first { +modify-file+ } _ queue-change ] each ;
+    '[
+        first2 V{ } clone swap {
+            [ kFSEventStreamEventFlagItemCreated bitand zero? [ +add-file+ suffix! ] unless ]
+            [ kFSEventStreamEventFlagItemRemoved bitand zero? [ +remove-file+ suffix! ] unless ]
+            [ kFSEventStreamEventFlagItemRenamed bitand zero? [ +rename-file+ suffix! ] unless ]
+        } cleave [ { +modify-file+ } ] [ >array ] if-empty _ queue-change
+    ] each ;
 
 M:: macosx (monitor) ( path recursive? mailbox -- monitor )
     path normalize-path :> path
     path mailbox macosx-monitor new-monitor
     dup [ enqueue-notifications ] curry
-    path 1array 0 0 <event-stream> >>handle ;
+    path 1array 0 kFSEventStreamCreateFlagFileEvents <event-stream> >>handle ;
 
 M: macosx-monitor dispose*
     [ handle>> dispose ] [ call-next-method ] bi ;
```

**Comment by erg** [comment]
This patch crashes for me on ``"io.monitors" test`` with an ``already disposed`` popup.

**Comment by mrjbq7** [comment]
Fixed in fc74498a9c4b6b1e7570f1d903272348be4bc05f
